### DISCUSSION TITLE
New cheap Sensirion CO2 I2C Sensor

### DISCUSSION THREAD

**Discussion by Mattes-in-Labratory** [arendst/Tasmota#11329] [Ideas]
It would be great to habe the new Sensirion SCD41 I2C CO2 Sensor integrated in Tasmota. It is an affordable CO2 sensor, just released, a great replacement for MHZ19 with lower footprint and much higher precision CO2, temperature and humidity readings.

If anyone needs documentation - Sensirion published documentation, Arduino, Python and Raspberry examples on Github.
Unfortunately my capabilities are not enough to integrate it in Tasmota myself.
The big old expensive Brother SCD30 is already integrated in tasmota sensors....

**Comment by Frogmore42** [comment]
The SCD41 is slightly cheaper, but not really that much cheaper (looks like it is about $10 cheaper than the SCD30 for the bare part, but only $6 cheaper on a board, and more expensive as an evaluation kit).  It is physically smaller, which is a good thing.  It does not appear to be significantly better than the SCD30 in any way other than size.

**Reply by Frogmore42** [reply]
No parts or boards are available currently, only the evaluation kit is currently available.  So far, I am not convinced the is cheap enough to be worthwhile.  Maybe when it really is generally available and the price (for one) drops to the 60 min order quantity price, it will be worth looking at.  Till then, I will stick with the SCD30.

**Comment by sergios100** [comment]
Hi Frogmore,

I just got an SCD41 Sensor on an development board from digikey for round about 50$ here is the link:
[https://www.digikey.com/short/fb9m5pq8](url)

I tried myself by repleacing some commands in the already working SCD30 library of tasmota to get the sensor to run, but I did not suceed. The new sensor has the i2c Address of 0x62 and was found by the "i2cscan" command. But obviously just changing the some of the new commands is just not enough :-)

I got it to run on an arduino with the sample-Code from sensirion, as well as with the library of Sparkfun 
[http://librarymanager/All#SparkFun_SCD4x](url)

The Datasheet from Sensirion with the commands reference can be found here:
[https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/9.5_CO2/Sensirion_CO2_Sensors_SCD4x_Datasheet.pdf](url)

Would be great if you could have a look at it! From my understanding this sensor uses no incandecent "heat source" and should be able to run much longer before failure.

Best Regards!

sergios100

**Reply by Arnold-n** [reply]
Hi @Frogmore42 and sergios100,
I got the basic functionality for the SCD41 running (with some shortcuts for now), based on the SCD30 code (i.e., for my experiment: replacing the SCD30 code). What would be the best way to integrate this? The i2c address of SCD30 and SCD4x are different so they can be used in parallel; hence a separate driver (lib and file) would perhaps be the easiest way to support SCD30 and SCD4x in parallel?
Kind regards,
Arnold

**Reply by barbudor** [reply]
Hi
Different chip, better have a different driver. Even if they have similar functionality and won't necesseraly be used together.

**Reply by Arnold-n** [reply]
Hi barbudor,
Thanks, I'll start working on a version that is ready for review.

**Reply by sergios100** [reply]
Hi Arnold-n!

Were you able to create a version ready to test/review? If Yes, where can I find it?

Kind Regards!

sergios100

**Reply by Arnold-n** [reply]
Hi sergios100,
I am working on it, decided to implement full functionality (all functions). Will publish soon.
Kind regards,
Arnold

**Reply by Arnold-n** [reply]
Hi sergios100 and Frogmore42,
Thanks for your patience. The SCD40/41 support is now available for review in a draft pull request https://github.com/arendst/Tasmota/pull/13139. As I am new to this process, I am not sure how to proceed from here.
Kind regards,
Arnold

**Reply by sergios100** [reply]
Hi Arnold-n,

Great Job!

I just downloaded your pull request #13139 to a ESP8266 Dev. Board.

It's working fine, showing all the data on the Web Frontend:
![2021-09-15_15h50_46](https://user-images.githubusercontent.com/63123058/133450605-e09c49b5-c976-43cb-9522-1b761cc28c25.png)

I also tried to send commands to the sensor, for example: `SCD40Alt` but I just got `{"SCD40Alt":-1}` back:
![2021-09-15_16h19_34](https://user-images.githubusercontent.com/63123058/133451147-22d776f9-92b5-45fb-aa57-7571f1b04072.png)

Probably I'm doing something wrong? The only thing I did was to enable #define USE_SCD40 (Line 609) in my_user_config.h

If anyone had success, let me know how! :-)

Best Regards!

Sergio

**Reply by Arnold-n** [reply]
Hi sergios100,

Did you issue an SCD40Stop command (+500ms wait) first? Most commands can only be issued when no periodic (low power or regular) measurement is being done.

**Reply by sergios100** [reply]
Hi Arnold-n,

my bad, worked fine after I "stopped" the sensor first! 

![2021-09-15_21h50_10](https://user-images.githubusercontent.com/63123058/133499973-e21a33b6-d289-4026-918b-e6eeb0a156e4.png)


Perhaps something to note then in the Tasmota Wiki as other users might run into the same issue? Oh and by the way: note as well, that to "re-start" the sensor the command is `SCD40Strt` as I had to dig into `tasmota/xsns_92_scd40.ino` to find the appropriate command :-)

I compiled and tested the Software on an ESP32 as well, works great there too!

Best Regards!

Sergio

**Reply by Arnold-n** [reply]
Hi sergios100,
Great, thanks for testing it on ESP32 too!
I'll start working on the documentation, for now it's only documented in the code or datasheet, but it is also not reviewed or approved for merge yet.
